### PR TITLE
refactor:  Connectivity Documentation

### PR DESCRIPTION
**Pull request by Johannes-Schneider**
## What Has Changed?

SAP/cloud-sdk-java-backlog/issues/179

This PR reworks our connectivity documentation so it is ready for the upcoming release of the SAP Cloud SDK for Java version 5 release.

### CODE REVIEW

File: docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+```
+
+It is **strongly recommended** to use the shown code **every time** a destination is needed as [explained below](#how-authentication-at-the-target-system-works).
+
+## About the `DestinationLoader` API
+
+Diving deeper into the `DestinationAccessor` reveals that it is actually a facade for the `DestinationLoader` API.
+
+A `DestinationLoader` is responsible for loading destinations, identified by their name, from a specific source.
+The SAP Cloud SDK comes with a number of `DestinationLoader` implementations that can be used to load destination from, for example, the [Cloud Foundry Destination Service](#the-destinationservice) or the [application's environment variables](#the-envvardestinationloader).
```
Comment (newtork) [inline docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx:46]: (Question)

Wait I thought SAP owned the destination service, not Cloud Foundry(?)

File: docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+As such, it can be used to retrieve both individual destinations as well as all destinstions stored for either a subscriber or the provider tenant.
+
+### How Authentication at the Target System Works
+
+The [Destination Service on Cloud Foundry](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) offers a [broad range of authentication methods](https://help.sap.com/docs/connectivity/sap-btp-connectivity-cf/http-destinations#configuring-authentication).
+
+When retrieving any destination, the Cloud Foundry Destination Service takes over the heavy lifting of authenticating at the target system on behalf of the calling application.
+Once authentication has been performed successfully, the Cloud Foundry Destination Service returns a destination configuration that contains all the information required to connect to the target system - including, for example, an `Authorization` header that can be used by the application.
+
+As a consequence, the SAP Cloud SDK's `DestinationService` abstraction automatically supports all authentication methods offered by the Cloud Foundry Destination Service.
```
Comment (newtork) [inline docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx:62]: (Question)

Should we point out that target-system-related token-exchanges are no (longer) happening in application runtime?
Just recently I had to tell that to a user on SNOW.

File: docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  - accessor
+  - destinationaccessor
+  - java
+  - connectivity
+  - btp
+---
+
+# About the `DestinationAccessor` API
+
+The `DestinationAccessor` API is part of the [`cloudplatform-connectivity` dependency](https://search.maven.org/artifact/com.sap.cloud.sdk.cloudplatform/cloudplatform-connectivity) and provides easy and convenient access to destinations that can be identified by their name.
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx:22]: Personally, when I'm reading some documentation, I'm always annoyed if the devs don't mention which dependency is needed for that certain class (or piece of functionality) I want to use.

But we could certainly use the proper `<name>` of the modules instead of the (more technical) `<artifactId>` - personally, when I include dependencies, I couldn't care less about what's in the `<name>`. All I need is the `<groupId>` and `<artifactId>`, but no strong opinion.

---

> Also I'm always struggling with the distinction of Maven terms: module, dependency, artifact.

I also struggled with this one. I'll read it up again and then use the (hopefully 🤞) correct term.

[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java_versioned_docs/version-v5/features/connectivity/destination/destination-accessor.mdx]: I chose _artifact_ in accordance with [the Maven Glossary](https://maven.apache.org/glossary.html).

File: docs-java/features/connectivity/destination/destination-accessor.mdx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+import io.vavr.control.Try;
+
+Try<Destination> maybeMyDestination = DestinationAccessor.tryGetDestination("my-destination");
+```
+
+It is **strongly recommended** to use the shown code **every time** a destination is needed as [explained below](#how-authentication-at-the-target-system-works).
+
+## About the `DestinationLoader` API
+
+Diving deeper into the `DestinationAccessor` reveals that it is actually a facade for the [`DestinationLoader`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationLoader.html) API.
```
Comment (MatKuhr) [inline docs-java/features/connectivity/destination/destination-accessor.mdx:43]: I feel like this information could be buried further down. Most readers will be interested in "how can I use a destination from BTP Cockpit" or "how can I add a header to my destination" or "how can I test locally". 

Almost nobody will need to know that we have the `DestinationLoader` at all.

**Review by MatKuhr — COMMENTED** [review]
I am a bit confused. Are the docs intended as addition to the existing files? If so, I would argue we should remove the existing docs and merge them into the new files. And the quality of the existing docs is just not good enough IMO.

Instead, if the old files are only here for easier comparison, then I'm not so sure about the structure and content of the new files. Right now there is quite some information from the old files missing in the new ones. Also, the new ones are structured along the lines of our API classes. This structure is somewhat flawed, as arguably the DestinationAccessor should comprise all destinations and really be just the accessor API. Thus I don't think it's easy for users to navigate our features by this API distinction.

**Comment by Johannes-Schneider** [comment]
@MatKuhr wrote:

> I am a bit confused. Are the docs intended as addition to the existing files? [ ... ] 

No, the old file is kept (for the moment) just for comparison.
The intention is indeed to replace all of our existing docs.

---

> Also, the new ones are structured along the lines of our API classes. This structure is somewhat flawed, as arguably the DestinationAccessor should comprise all destinations and really be just the accessor API.

I do agree that our current API structure is somewhat incomplete / all over the place.
Nevertheless, I'd argue that it doesn't make sense to document anything but the current state.

Of course, if we somehow unify the `DestinationAccessor` and `ServiceBindingDestinationAccessor`, we have to update this documentation to reflect those changes then as well.

We can of course still discuss whether we want to make the `DestinationAccessor` more prominent as it is most probably the API most users would want to interact with (e.g. we could put a code sample to the Overview page).

---

> Thus I don't think it's easy for users to navigate our features by this API distinction.

Do you have any suggestions about how it would become easier?

My idea was to outline **where** destinations might be coming from in the first place (Overview page) and then (in the detail pages) explain how exactly to use the APIs and how they work internally.
I thought customers should be able to quickly identify the correct documentation as long as they know what sort of Destinations they are interested in.
At the same time, customers who just want to know what kind of Destination sources the SDK supports would also get an overview quickly without being overwhelmed by code samples and implementation details.

Furthermore, I tried structuring the detail pages in a top-down fashion so that the most important things (i.e. just API usage) can be found quickly without the need to dig through too many samples.

**Review by MatKuhr — COMMENTED** [review]
partial review

File: docs-java/features/connectivity/004-http-destinations.mdx [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+6. (Conditional) Headers that are dervied from [authentication related properties](#deriving-the-authorization-header), **iff** no `Authorization` header was provided by any of the previous options.
+7. (Conditional) Headers that are derived from [proxy authentication related properties](#deriving-the-proxy-authorization-header), **iff** no `Proxy-Authorization` header was provided by any of the previous options.
+
+### On-Premise Handling
+
+The `DefaultHttpDestination` implements dedicated support for on-premise connectivity as [documented here](on-premise).
+
+### Deriving The `Authorization` Header
+
+As [explained above](#about-http-headers), the `DefaultHttpDestination` tries to derive the `Authorization` header from authentication related properties, **iff** no `Authorization` header is provided by any other source.
```
Comment (MatKuhr) [inline docs-java/features/connectivity/004-http-destinations.mdx:58]:
```suggestion
The `DefaultHttpDestination` tries to derive the `Authorization` header from authentication related properties, **iff** no `Authorization` header is provided by any other source.
```

(minor) I feel like there is not much value in referencing text that is 3-4 sentences away above 😉

Comment (MatKuhr) [reply docs-java/features/connectivity/004-http-destinations.mdx:58]: (minor) Also, this would be a good place to mention those typically come from the BTP destination service, linking to that doc?

File: docs-java/features/connectivity/004-http-destinations.mdx [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+:::
+
+The [`DefaultHttpDestination.Builder`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html) allows you to create a `DefaultHttpDestination` instance from scratch.
+It allows you to set any desired property, header, header provider, or any other value that is supported by the `DefaultHttpDestination` implementation.
+
+### Building On Top Of An Existing Destination
+
+Sometimes, you may want to slightly adjust an existing `Destination` instance.
+For example, there might be just this one single header missing for a specific request. <br />
+Unfortunately, adding this header directly to the existing `Destination` instance is not possible, as [the `Destination` interface is immutable](#about-immutability).
```
Comment (MatKuhr) [inline docs-java/features/connectivity/004-http-destinations.mdx:110]: (minor) Personally, I would omit terms like unfortunately / luckily to keep the type of language consistent

File: docs-java/features/connectivity/004-http-destinations.mdx [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+It will **not** override any existing header with the same name.
+
+### Adding Individual Header Providers
+
+The [`headerProviders(DestinationHeaderProvider...)`](<pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DefaultHttpDestination.Builder.html#headerProviders(com.sap.cloud.sdk.cloudplatform.connectivity.DestinationHeaderProvider...)>) method allows you to add as many `DestinationHeaderProvider`s to the `DefaultHttpDestination` instance as you wish.
+Once again, existing header providers will **not** be overridden.
+
+### Adding Static Header Providers
+
+In some scenarios, there are certain HTTP headers that should always be sent to a destination, regardless of the request.
```
Comment (MatKuhr) [inline docs-java/features/connectivity/004-http-destinations.mdx:166]: (minor) Technically this is unrelated to the Builder API. Since we have a dedicated section about header providers below, maybe just move it there?

**Review by MatKuhr — COMMENTED** [review]
more review comments 🥲

But the overall state looks really good already 🥳

File: docs-java/features/connectivity/000-overview.mdx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+- From [Service Bindings](https://github.com/SAP/btp-environment-variable-access/wiki/Fundamentals#service-binding).
+  - Via the [`ServiceBindingDestinationLoader` API](service-bindings)
+- From existing `Destination` objects.
+  - Via the [`DefaultHttpDestination` API](http-destinations)
+
+## Using Destinations
+
+Once a [`Destination`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/Destination.html) has been retrieved, it can be used to connect to the system or service it represents.
+This is done by converting the given `Destination` into an HTTP client, which is then used to send requests to the system or service.
+
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/000-overview.mdx:48]: I don't feel like reading destination properties belongs on the "Connectivity Overview" page as it isn't related to Cloud SDK connectivity features from my point of view.
Moreover, as reading the destination properties only makes sense when destinations are retrieved from the BTP Destination service, putting this functionality so prominently on our overview page feels odd to me.

But adding a section on the BTP Destination Service integration pages makes sense I think 👍

File: docs-java/features/connectivity/001-btp-destination-service.mdx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+```
+
+Note that you need an actual instance of `DestinationLoader` (e.g. `DestinationService`) to be able to supply the `DestinationOptions`.
+The `DestinationAccessor` **does not** offer this API.
+
+</details>
+
+### Resilience
+
+Since the `DestinationService` is performing error prone operations (i.e. HTTP calls), it comes with some resilience mechanisms out of the box:
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/001-btp-destination-service.mdx:130]: I think that would be a great feature. Unfortunately, the [corresponding constructor is package private](https://github.com/SAP/cloud-sdk-java/blob/main/cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java#L54).

Therefore, customers cannot change the resilience behavior.

File: docs-java/features/connectivity/001-btp-destination-service.mdx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+This is to prevent excessive memory usage in case of a large number of tenants and/or destinations.
+
+### Tweaking the Caching Behavior
+
+Even though the [default caching behavior](#default-caching-behavior) should serve as a good starting point for many applications, it may be necessary to tweak it for specific use cases.
+For that, the [`DestinationService.Cache`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.Cache.html) offers a number of configuration APIs ([JavaDoc](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.Cache.html)).
+
+#### Caution
+
+When using any of these APIs, the underlying cache instance will be **recreated**.
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/001-btp-destination-service.mdx:158]: > the Cloud SDK will log a warning

That's true. But who is checking for logs if everything seems to be working?

File: docs-java/features/connectivity/001-btp-destination-service.mdx (hunk 9 above)
[the remark this answers is not in the document]
Comment (MatKuhr) [reply docs-java/features/connectivity/001-btp-destination-service.mdx:130]: Of course it is private, otherwise one might use that over the [builder](https://github.com/SAP/cloud-sdk-java/blob/main/cloudplatform/connectivity-destination-service/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/DestinationService.java#L812) we have for it 😉

[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/001-btp-destination-service.mdx:167]: Of course we have a builder .... 🤦‍♂️😅

File: docs-java/features/connectivity/001-btp-destination-service.mdx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+Since the `DestinationService` is performing error prone operations (i.e. HTTP calls), it comes with some resilience mechanisms out of the box:
+
+- (**request isolated**) _Time Limiter_ - Destination retrieval is limited to a maximum of **6 seconds**.
+- (**tenant isolated**) _Circuit Breaker_ - Requests will be blocked once more than **50%** of the **last 10** requests failed.
+  - For fetching **individual** destinations, the circuit breaker will stay open for **6 seconds**.
+  - For fetching **all** destinations, the circuit breaker will stay open for **60 seconds**.
+- (**tenant isolated**) _Bulk Head_ - The maximum number of concurrent requests is limited to **50**. Queued tasks will wait for up to **60 seconds** before being rejected.
+
+### Default Caching Behavior
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/001-btp-destination-service.mdx]: Yeah... the headings (h2 vs. h3) are somewhat odd right now. I changed a few of them 🤷‍♂️

File: docs-java/features/connectivity/004-http-destinations.mdx [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+One of the most important properties of any `HttpDestination` are the HTTP headers, which are attached to any outgoing request performed with the destination.
+
+The `DefaultHttpDestination` implements support for following sources of HTTP headers:
+
+1. Properties named using the pattern `URL.header.<header-name>` will be converted into HTTP headers.
+2. [Special properties](#special-header-providers), which will be converted into HTTP headers.
+3. Headers that were [directly added](#adding-individual-headers) via the [Builder API](#about-the-builder-api) while assembling the `DefaultHttpDestination` instance.
+4. Headers that are provided by [`DestinationHeaderProvider`s](#about-headerproviders).
+5. Any headers required for on-premise connectivity will automatically be added, **iff** the `ProxyType` property is set to `OnPremise` as [explained below](#on-premise-handling).
```
Comment (KavithaSiva) [inline docs-java/features/connectivity/004-http-destinations.mdx:48]:
```suggestion
5. Any headers required for on-premise connectivity will automatically be added, **iff** the `ProxyType` property is set to `OnPremise` as [explained below](#deriving-the-proxy-authorization-header).
```

File: docs-java/features/connectivity/004-http-destinations.mdx [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+- For `AuthenticationType.BASIC_AUTHENTICATION`, the basic credentials are taken from the `User` (or `username`, if the former doesn't exist) and the `Password` properties. <br />
+  If these properties are not set, an exception is thrown.
+
+- For `AuthenticationType.TOKEN_FORWARDING`, the `RequestHeaderAccessor` is used to find any `Authorization` header in the current request. All headers that are found (there should usually only be one) are then forwarded.
+
+### Deriving The `Proxy-Authorization` Header
+
+:::info On-Premise Handling
+
+As described [above](#on-premise-handling), the `DefaultHttpDestination` implements dedicated support for on-premise connectivity.
```
Comment (KavithaSiva) [inline docs-java/features/connectivity/004-http-destinations.mdx:69]:
```suggestion
As described above, the `DefaultHttpDestination` implements dedicated support for on-premise connectivity.
```

Current link is dead.

File: docs-java/features/connectivity/003-service-bindings.mdx [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+The SAP Cloud SDK, therefore, provides a way to convert [`ServiceBinding`](https://github.com/SAP/btp-environment-variable-access/blob/main/api-parent/core-api/src/main/java/com/sap/cloud/environment/servicebinding/api/ServiceBinding.java) instances into `HttpDestination` instances.
+
+```java
+import com.sap.cloud.environment.servicebinding.api.ServiceIdentifier;
+
+var options = ServiceBindingDestinationOptions.forService(ServiceIdentifier.DESTINATION).build();
+HttpDestination destination = ServiceBindingDestinationLoader.defaultLoaderChain().getDestination(options);
+```
+
+## How It Works
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/003-service-bindings.mdx:33]: I'd agree with the multitenancy.

However, I'm not so sure about caching... Since our OAuth loader already caches auth tokens across `Destination` instances, it doesn't really matter whether the `Destination` objects are cached (since creating them is not too costly).

File: docs-java/features/connectivity/003-service-bindings.mdx [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+    API2;
+
+    @Nonnull
+    @Override
+    public MyApiChoice getValue() {
+        return this;
+    }
+}
+
+class MyHttpHeader implements ServiceBindingDestinationOptions.OptionsEnhancer<String> {
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/003-service-bindings.mdx:95]: Agreed. I changed that.

File: docs-java/features/connectivity/003-service-bindings.mdx [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+## About the _OAuth2Loader_
+
+The [`OAuth2ServiceBindingDestinationLoader`](pathname:///java-api/v5/com/sap/cloud/sdk/cloudplatform/connectivity/OAuth2ServiceBindingDestinationLoader.html) comes as part of the [`connectivity-oauth` artifact](https://search.maven.org/artifact/com.sap.cloud.sdk.cloudplatform/connectivity-oauth) and provides a customizable implementation for converting `ServiceBinding`s that use OAuth2 authentication into `HttpDestination` instances. <br />
+More precisely, it supports the **OAuth2 Client Credentials** and **OAuth2 Client Certificate** grant types.
+
+Internally, the implementation performs the following steps:
+
+1. Check whether the given `ServiceBindingDestinationOptions` can be transformed (e.g. whether the `ServiceBinding` uses a supported grant type).
+2. Extract the required information from the `ServiceBinding` (i.e. the service URL, the token URL, and the client identity).
+3. Check whether the `ServiceBindingDestinationOptions` contain another `HttpDestination` that should be _proxied_.
```
[the remark this answers is not in the document]
Comment (Johannes-Schneider) [reply docs-java/features/connectivity/003-service-bindings.mdx]: That's a fair point.

I figured to simply remove the documentation about proxied destinations, as this is an internal feature only.

**Review by KavithaSiva — APPROVED** [review]
LGTM.
